### PR TITLE
On result2 resolution result have addresses or error

### PR DESCRIPTION
**Pull request by kannanjgithub**
Changes for passing the error status via ResolutionResult, and enhancing the ResolutionResult for the same. This PR is branched from the branch used for https://github.com/grpc/grpc-java/pull/11313 that introduced onResult2 returning the Status.

### CODE REVIEW

File: api/src/main/java/io/grpc/StatusOr.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package io.grpc;
```
Comment (ejona86) [inline api/src/main/java/io/grpc/StatusOr.java:1]: Add copyright header.

File: api/src/main/java/io/grpc/StatusOr.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  public T value() {
+    return value;
+  }
+
+  public Status status() {
+    return status;
+  }
+
+  private Status status;
```
Comment (ejona86) [inline api/src/main/java/io/grpc/StatusOr.java:37]: Make these final. And yes, that means you need to pass them to a constructor.

Final is special in Java for synchronization. It allows the object to be used with unsafe publication. We don't really do that, but it is a helpful safety guarantee for immutable objects. For example, CallOptions was swapped to using final in https://github.com/grpc/grpc-java/commit/1b94f48dc84f91977b1553a33e8eccbfaffbfb05

File: api/src/main/java/io/grpc/StatusOr.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    checkArgument(!status.isOk(), "cannot use OK status: %s", status);
+    result.hasValue = false;
+    return result;
+  }
+
+  public boolean hasValue() {
+    return hasValue;
+  }
+
+  public T value() {
```
Comment (ejona86) [inline api/src/main/java/io/grpc/StatusOr.java:29]: Javadoc, and define what happens if there is no value.

File: api/src/main/java/io/grpc/StatusOr.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    return value;
+  }
+
+  public Status status() {
```
Comment (ejona86) [inline api/src/main/java/io/grpc/StatusOr.java:33]: Javadoc, and define what happens if there is a value. For example, maybe this returns an OK status.

File: api/src/main/java/io/grpc/StatusOr.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  public static <T> StatusOr<T> fromStatus(Status status) {
+    StatusOr<T> result = new StatusOr<T>();
+    result.status  = checkNotNull(status, "status");
+    checkArgument(!status.isOk(), "cannot use OK status: %s", status);
+    result.hasValue = false;
+    return result;
+  }
+
+  public boolean hasValue() {
+    return hasValue;
```
Comment (ejona86) [inline api/src/main/java/io/grpc/StatusOr.java:26]: Use `status == null` instead?

**Review by ejona86 — COMMENTED** [review]
Rebase this once #11313 goes in, as it is hard to read now. We'll want to discuss this in the next API review meeting.

**Comment by kannanjgithub** [comment]
I synced and resolved conflicts from the website, and ended up with some compilation errors. Will pull changes to my cloudtop, fix them and upload the changes.

File: api/src/main/java/io/grpc/StatusOr.java (hunk 5 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/StatusOr.java]: Done.

File: api/src/main/java/io/grpc/StatusOr.java (hunk 4 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/StatusOr.java:53]: Done.

File: api/src/main/java/io/grpc/StatusOr.java (hunk 2 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/StatusOr.java]: Done.

File: api/src/main/java/io/grpc/StatusOr.java (hunk 1 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/StatusOr.java:17]: Done.

**Comment by ejona86** [comment]
Let's discuss this in the new API review meeting.

File: api/src/main/java/io/grpc/StatusOr.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2015 The gRPC Authors
```
Comment (sergiitk) [inline api/src/main/java/io/grpc/StatusOr.java:2]:
```suggestion
 * Copyright 2024 The gRPC Authors
```

 😺

**Comment by kannanjgithub** [comment]
> Let's discuss this in the next API review meeting.

Implemented changes discussed:
1. Equals/hashCode
2. of() instead of fromValue()
3. getValue()/getStatus() instead of value()/status().

File: api/src/main/java/io/grpc/StatusOr.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    return value != null ? Status.OK : status;
+  }
+
+  @Override
+  @SuppressWarnings({"ReferenceEquality", "unchecked"})
+  public boolean equals(Object other) {
+    if (!(other instanceof StatusOr)) {
+      return false;
+    }
+    StatusOr<T> otherStatus = (StatusOr<T>) other;
```
Comment (ejona86) [inline api/src/main/java/io/grpc/StatusOr.java:73]: I think the `unchecked` warning was on to something here. We don't know the types are the same. This should be `StatusOr<?>` instead.

File: api/src/main/java/io/grpc/StatusOr.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    if (hasValue() != otherStatus.hasValue()) {
+      return false;
+    }
+    if (hasValue()) {
+      return Objects.equal(value, otherStatus.value);
+    }
+    return status == otherStatus.status;
```
Comment (ejona86) [inline api/src/main/java/io/grpc/StatusOr.java:80]: I agree this works, but let's avoid the ReferenceEquality warning and use equals() or Objects.equal() here.

File: api/src/main/java/io/grpc/StatusOr.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+  public @Nullable T getValue() {
+    if (status != null) {
+      throw new IllegalStateException("No value present.");
+    }
+    return value;
+  }
+
+  /** Returns the status. If there is a value, returns OK. */
+  public Status getStatus() {
+    return value != null ? Status.OK : status;
```
Comment (ejona86) [inline api/src/main/java/io/grpc/StatusOr.java:64]: This needs to check `status == null`, not `value`, since now value can be null.

File: api/src/main/java/io/grpc/NameResolver.java [hunk 10]
```diff
@@ -699,8 +723,20 @@ public static final class Builder {
...
+        return this;
+      }
+
+      /**
+       * Sets the addresses resolved by name resolution or the error in doing so. This field is
+       * required.
+       * @param addresses Resolved addresses or an error in resolving addresses
+       */
+      public Builder setAddressesOrError(StatusOr<List<EquivalentAddressGroup>> addresses) {
         this.addresses = addresses;
```
Comment (ejona86) [inline api/src/main/java/io/grpc/NameResolver.java:740]: Add `checkNotNull(addresses)` here.

Hint: for the deprecated method have it call the new method, even though it is trivial to inline, simply so that we don't have to think about whether it also needs the checks.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 11]
```diff
@@ -1693,18 +1694,24 @@ public Status onResult2(final ResolutionResult resolutionResult) {
...
-          "Resolved address: {0}, config={1}",
-          servers,
-          resolutionResult.getAttributes());
-
-      if (lastResolutionState != ResolutionState.SUCCESS) {
-        channelLogger.log(ChannelLogLevel.INFO, "Address resolved: {0}", servers);
-        lastResolutionState = ResolutionState.SUCCESS;
+      StatusOr<List<EquivalentAddressGroup>> serversOrError =
+          resolutionResult.getAddressesOrError();
+      if (serversOrError != null && serversOrError.hasValue()) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:1699]: I suggested above to make serversOrError guaranteed non-null. Things like this become simpler.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 12]
```diff
@@ -1806,12 +1813,17 @@ public Status onResult2(final ResolutionResult resolutionResult) {
...
         Attributes attributes = attrBuilder.build();
 
-        return helper.lb.tryAcceptResolvedAddresses(
-            ResolvedAddresses.newBuilder()
-                .setAddresses(servers)
-                .setAttributes(attributes)
-                .setLoadBalancingPolicyConfig(effectiveServiceConfig.getLoadBalancingConfig())
-                .build());
+        ResolvedAddresses.Builder resolvedAddresses = ResolvedAddresses.newBuilder();
+        if (serversOrError != null && serversOrError.hasValue()) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:1817]: It looks like this is currently guaranteed to be `true`, because of the early return.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 13]
```diff
@@ -1693,18 +1694,24 @@ public Status onResult2(final ResolutionResult resolutionResult) {
...
+            "Resolved address: {0}, config={1}",
+            serversOrError,
+            resolutionResult.getAttributes());
+
+        if (lastResolutionState != ResolutionState.SUCCESS) {
+          channelLogger.log(ChannelLogLevel.INFO, "Address resolved: {0}",
+              serversOrError.getValue());
+          lastResolutionState = ResolutionState.SUCCESS;
+        }
+      } else if (serversOrError != null && !serversOrError.hasValue()) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:1711]: Since this block returns, you may consider making it separate early return (not if/else):
```java
if (!serversOrError.hasValue()) {
  handleErrorInSyncContext(serversOrError.getStatus());
  return serversOrError.getStatus();
}

List<EquivalentAddressGroup> servers = serversOrError.getValue();
// the rest of the previous code
```

**Review by ejona86 — APPROVED** [review]

File: api/src/main/java/io/grpc/StatusOr.java (hunk 8 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/StatusOr.java]: Done.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 11 above)
Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Done.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 12 above)
Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Done.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 13 above)
Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Done.

**Comment by kannanjgithub** [comment]
Code coverage check was failing including for some of the existing lines that I touched that previously didn't have coverage. Fixed all code coverage errors. PTAL @ejona86 .

File: netty/src/main/java/io/grpc/netty/UdsNameResolver.java [hunk 14]
```diff
@@ -57,8 +60,9 @@ private void resolve() {
     ResolutionResult.Builder resolutionResultBuilder = ResolutionResult.newBuilder();
     List<EquivalentAddressGroup> servers = new ArrayList<>(1);
     servers.add(new EquivalentAddressGroup(new DomainSocketAddress(authority)));
-    resolutionResultBuilder.setAddresses(Collections.unmodifiableList(servers));
-    listener.onResult(resolutionResultBuilder.build());
+    resolutionResultBuilder.setAddressesOrError(StatusOr.fromValue(servers));
+    syncContext.execute(() ->
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/UdsNameResolver.java:64]: Why this change? start and refresh must be called from the synchronization context.

File: api/src/test/java/io/grpc/NameResolverTest.java [hunk 15]
```diff
@@ -17,20 +17,42 @@
...
 import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
 
 /** Unit tests for the inner classes in {@link NameResolver}. */
 @RunWith(JUnit4.class)
 public class NameResolverTest {
+  @Rule
+  public final MockitoRule mocks = MockitoJUnit.rule();
```
Comment (ejona86) [inline api/src/test/java/io/grpc/NameResolverTest.java:48]: Let's not mix this among the `static` fields. Can you move this down so it is with the other non-static fields?

File: api/src/main/java/io/grpc/NameResolver.java [hunk 16]
```diff
@@ -652,11 +676,19 @@ public ConfigOrError getServiceConfig() {
 
     @Override
     public String toString() {
-      return MoreObjects.toStringHelper(this)
-          .add("addresses", addresses)
-          .add("attributes", attributes)
-          .add("serviceConfig", serviceConfig)
-          .toString();
+      ToStringHelper stringHelper = MoreObjects.toStringHelper(this);
+      if (addressesOrError.hasValue()) {
```
Comment (ejona86) [inline api/src/main/java/io/grpc/NameResolver.java:680]: Seems we should add a `toString()` to StatusOr.

File: api/src/main/java/io/grpc/NameResolver.java [hunk 17]
```diff
@@ -584,17 +587,27 @@ public abstract static class ServiceConfigParser {
...
+    ResolutionResult(
+        StatusOr<List<EquivalentAddressGroup>> addressesOrError,
+        @ResolutionResultAttr Attributes attributes,
+        ConfigOrError serviceConfig) {
+      this.addressesOrError = addressesOrError;
+      this.attributes = checkNotNull(attributes, "attributes");
+      this.serviceConfig = serviceConfig;
+    }
+
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/NameResolver.java]: Done.

File: api/src/main/java/io/grpc/NameResolver.java (hunk 16 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/NameResolver.java]: Done.

File: api/src/test/java/io/grpc/NameResolverTest.java (hunk 15 above)
Comment (kannanjgithub) [reply api/src/test/java/io/grpc/NameResolverTest.java]: Done.

File: netty/src/main/java/io/grpc/netty/UdsNameResolver.java (hunk 14 above)
Comment (kannanjgithub) [reply netty/src/main/java/io/grpc/netty/UdsNameResolver.java]: Done.

**Review by ejona86 — APPROVED** [review]
